compiler/rustc_infer/src/infer/error_reporting/note_and_explain.rs

```diff
@@ -78,48 +78,82 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                         if tcx.def_kind(proj.def_id) != DefKind::ImplTraitPlaceholder =>
                     {
                         let generics = tcx.generics_of(body_owner_def_id);
-                        let p_span = tcx.def_span(generics.type_param(p, tcx).def_id);
+                        let p_def_id = generics.type_param(p, tcx).def_id;
+                        let p_span = tcx.def_span(p_def_id);
                         if !sp.contains(p_span) {
                             diag.span_label(p_span, "this type parameter");
                         }
                         let hir = tcx.hir();
                         let mut note = true;
-                        if let Some(generics) = generics
-                            .type_param(p, tcx)
-                            .def_id
+                        let parent = p_def_id
                             .as_local()
-                            .map(|id| hir.local_def_id_to_hir_id(id))
-                            .and_then(|id| tcx.hir().find_parent(id))
-                            .as_ref()
-                            .and_then(|node| node.generics())
+                            .map(|id| (id, hir.local_def_id_to_hir_id(id)))
+                            .and_then(|(local_id, hir_id)| tcx.hir().find_parent(hir_id).map(|node| (local_id, node)));
+                        if let Some((local_id, generics)) = parent.as_ref()
+                            .and_then(|(local_id, node)| node.generics().map(|generics| (local_id, generics)))
```

Comment:
There's gotta be a way to simplify this, perhaps by avoiding these combinators... Could you maybe make this logic a bit simpler?

Reply:
Couldn't get rid of an `and_then`, but does this look cleaner to you?